### PR TITLE
refactor: persistent kvstore abci

### PR DESCRIPTION
**Pull request by shotonoff**
## Issue being fixed or feature implemented

When refactoring the `persistent-kvstore` application, I had the following goals:
* remove unused (or deprecated)  logic
* change API to be able to reuse the functions out of the app
* use the new API where possible

## What was done?

* refactored `persistent-kvstore`, the most significant change is a replaced of deprecated handlers for set/update a specific part of ValidatorSetUpdate to set the whole entity
* added implementation of `LLMQToValidatorSetProto` to convert LLMQ generated data into protobuf ValidatorSetUpdate structure
* added MarshalValidatorSetUpdate/UnmarshalValidatorSetUpdate functions
* fixed the affected tests by this refactoring

## How Has This Been Tested?

Unit/e2e tests

## Breaking Changes

All changes are non-critical and are used only in tests

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

File: internal/state/state_test.go [hunk 1]
```diff
@@ -858,80 +830,21 @@ func TestFourAddFourMinusOneGenesisValidators(t *testing.T) {
...
-			EndBlock:   &abci.ResponseEndBlock{ValidatorSetUpdate: &abciValidatorSetUpdate},
-		}
-		block, err := statefactory.MakeBlock(oldState, oldState.LastBlockHeight+1, new(types.Commit), nil, 0)
-		require.NoError(t, err)
-		blockID := types.BlockID{Hash: block.Hash(), PartSetHeader: block.MakePartSet(testPartSize).Header()}
-		state, err = sm.UpdateState(state, firstNodeProTxHash, blockID, &block.Header, abciResponses,
-			validatorUpdates, thresholdPublicKey3, quorumHash)
+		abciValidatorSetUpdate, err := abci.LLMQToValidatorSetProto(*ld)
 		require.NoError(t, err)
+		abciValidatorSetUpdate.QuorumHash = quorumHash
```
Comment (lklimek) [inline internal/state/state_test.go:835]: Not sure, but shouldn't we regenerate quorum hash when we add/remove validator set members?

File: internal/state/state_test.go [hunk 2]
```diff
@@ -858,80 +830,21 @@ func TestFourAddFourMinusOneGenesisValidators(t *testing.T) {
...
-		abciResponses := &tmstate.ABCIResponses{
-			BeginBlock: &abci.ResponseBeginBlock{},
-			EndBlock:   &abci.ResponseEndBlock{ValidatorSetUpdate: &abciValidatorSetUpdate},
-		}
-		block, err := statefactory.MakeBlock(oldState, oldState.LastBlockHeight+1, new(types.Commit), nil, 0)
-		require.NoError(t, err)
-		blockID := types.BlockID{Hash: block.Hash(), PartSetHeader: block.MakePartSet(testPartSize).Header()}
-		state, err = sm.UpdateState(state, firstNodeProTxHash, blockID, &block.Header, abciResponses,
-			validatorUpdates, thresholdPublicKey3, quorumHash)
+		abciValidatorSetUpdate, err := abci.LLMQToValidatorSetProto(*ld)
```
Comment (lklimek) [inline internal/state/state_test.go:833]: this adds validators one by one, we should also test change of multiple validators at once

**Review by lklimek — COMMENTED** [review]
This code is soo muuch better :) 

Left some comments.

Also, see the linter output. 

Please don't merge before #310

File: abci/example/kvstore/persistent_kvstore.go [hunk 3]
```diff
@@ -81,7 +63,7 @@ func (app *PersistentKVStoreApplication) DeliverTx(req types.RequestDeliverTx) t
 	app.mtx.Lock()
 	defer app.mtx.Unlock()
 	if isValidatorSetUpdateTx(req.Tx) {
-		err := app.execValidatorSet(req.Tx)
+		err := app.execValidatorSetTx(req.Tx)
 		if err != nil {
 			return types.ResponseDeliverTx{
 				Code: code.CodeTypeEncodingError,
```
[the remark this answers is not in the document]
Comment (shotonoff) [reply abci/example/kvstore/persistent_kvstore.go]: changed on `CodeTypeUnknownError` but it is still un precise in my opinion, 
the actual list is following and it looks poor:
```
CodeTypeOK
CodeTypeEncodingError
CodeTypeBadNonce
CodeTypeUnauthorized
CodeTypeUnknownError
```

File: abci/example/kvstore/persistent_kvstore.go [hunk 4]
```diff
@@ -97,68 +79,48 @@ func (app *PersistentKVStoreApplication) CheckTx(req types.RequestCheckTx) types
...
-		return resQuery
-	case "/val":
-		key := []byte(valSetTxKey(string(reqQuery.Data)))
-		value, err := app.app.state.db.Get(key)
+	case "/vsu":
+		vsu, err := app.repo.get()
 		if err != nil {
-			panic(err)
+			return types.ResponseQuery{
+				Code: code.CodeTypeEncodingError,
```
[the remark this answers is not in the document]
Comment (shotonoff) [reply abci/example/kvstore/persistent_kvstore.go]: changed, details [here](https://github.com/dashevo/tenderdash/pull/313/files#r837443751)

**Review by shotonoff — COMMENTED** [review]
please have a look at the modifications / answers

File: internal/state/state_test.go [hunk 5]
```diff
@@ -774,10 +774,26 @@ func TestFourAddFourMinusOneGenesisValidators(t *testing.T) {
...
+		resp := &tmstate.ABCIResponses{
+			BeginBlock: &abci.ResponseBeginBlock{},
+			EndBlock:   &abci.ResponseEndBlock{ValidatorSetUpdate: vsu},
+		}
+		validatorUpdates, thresholdPubKey, quorumHash, err :=
+			types.PB2TM.ValidatorUpdatesFromValidatorSet(resp.EndBlock.ValidatorSetUpdate)
+		require.NoError(t, err)
+		block, err := statefactory.MakeBlock(prevState, prevState.LastBlockHeight+1, new(types.Commit), nil, 0)
+		require.NoError(t, err)
+		blockID := types.BlockID{Hash: block.Hash(), PartSetHeader: block.MakePartSet(testPartSize).Header()}
```
[the remark this answers is not in the document]
Comment (lklimek) [reply internal/state/state_test.go]: I mean to add in types/block.go method  like:

```
func (block * Block) BlockID() BlockID {
return BlockID{Hash: block.Hash(), PartSetHeader: block.MakePartSet(testPartSize).Header()}
}
```

File: internal/state/state_test.go [hunk 6]
```diff
@@ -778,25 +778,12 @@ func TestFourAddFourMinusOneGenesisValidators(t *testing.T) {
...
-		require.NoError(t, err)
-		block, err := statefactory.MakeBlock(prevState, prevState.LastBlockHeight+1, new(types.Commit), nil, 0)
-		require.NoError(t, err)
-		blockID := types.BlockID{Hash: block.Hash(), PartSetHeader: block.MakePartSet(testPartSize).Header()}
-		state, err = sm.UpdateState(state, firstProTxHash, blockID, &block.Header, resp,
-			validatorUpdates, thresholdPubKey, quorumHash)
-		require.NoError(t, err)
-		return state
-	}
+	execute := bockExecutorFunc(t, firstProTxHash)
```
Comment (lklimek) [inline internal/state/state_test.go:781]:
```suggestion
	execute := blockExecutorFunc(t, firstProTxHash)
```

File: internal/state/state_test.go [hunk 7]
```diff
@@ -1103,3 +1088,23 @@ func TestState_StateID(t *testing.T) {
 	err := stateID.ValidateBasic()
 	assert.NoError(t, err, "StateID validation failed")
 }
+
+func bockExecutorFunc(t *testing.T, firstProTxHash crypto.ProTxHash) func(prevState, state sm.State, vsu *abci.ValidatorSetUpdate) sm.State {
```
Comment (lklimek) [inline internal/state/state_test.go:1092]:
```suggestion
func blockExecutorFunc(t *testing.T, firstProTxHash crypto.ProTxHash) func(prevState, state sm.State, vsu *abci.ValidatorSetUpdate) sm.State {
```

**Review by lklimek — APPROVED** [review]

**Review by lklimek — APPROVED** [review]

**Pull request merged**